### PR TITLE
Add file preloading/caching 

### LINKED ISSUE #1329: [Request]: add vmtouch, ramdisk or other pre-caching support
### Tell us the problem or your need

With some older games the engines were badly optimized so there is a loading screen after every house, map or screen. I not sure how feasible it is with Bottles and Wine (it looks like vmtouch works with wine) but it would be nice to add some kind of feature that pre-loads the game into memory to cut down on long loading.

if I understand correctly then PlayOnLinux has vmtouch and I seen people talking about it:
https://askubuntu.com/questions/1104882/preload-for-playonlinux-application

### Describe the solution you'd like

Add vmtouch or ramdisk (via Wine?) or some other pre-memory or pre-caching loader where user can specify the game files to load into memory when game launches.

### Other solutions?

N/a

### Additional context and references

Of course this feature would not work for everyone because you need a lot of unused memory. My computer has 16gb and I don’t think it’s ever going above 12gb used so I might be able to take advantage. Those with 32gb definitely can take advantage.

### PR DESCRIPTION
**Pull request by axtloss**
# Description
This allows bottles to preload files into cache before starting the application, causing faster loading times in games and applications.

Fixes #1329 

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce.
- [ ] Test A
- [ ] Test B

---

Currently I only have the actual precaching implemented, freeing the cache after the application quits isn't yet implemented and the ui probably needs some tweaks too.

### CODE REVIEW

File: src/ui/details-preferences.ui [hunk 1]
```diff
@@ -572,6 +572,28 @@
                         </child>
                     </object>
                 </child>
+                <child>
+                  <object class="AdwActionRow">
+                    <property name="activatable-widget">switch_vmtouch</property>
+                    <property name="title" translatable="yes">Vmtouch</property>
+                    <property name="subtitle" translatable="yes">Preload game files into ram</property>
```
Comment (jannuary) [inline src/ui/details-preferences.ui:579]:
```suggestion
                    <property name="subtitle" translatable="yes">Preload game files into RAM</property>
```

**Comment by axtloss** [comment]
How would I add an option in the settings that allows setting a file size which always is in megabytes (M)?
Currently I just have an AdwEntry widget:
image

but I doubt that it's the right way to do it.

**Comment by mirkobrombin** [comment]
Can this setting be omitted? Maybe using a default?

**Comment by axtloss** [comment]
The default in vmtouch is 500M, but some users may want to cache bigger files, which is why that option would be needed.
Although I could hardcode a higher file size

**Comment by mirkobrombin** [comment]
I think we can set 1000 as standard and wait for user feedbacks without adding more options

**Comment by axtloss** [comment]
I removed the option, the pr should now be read to merge.

**Comment by jannuary** [comment]
I don't like how this expects users beforehand to 
1. Know whether or not they need to enable preloading
2. Know what files need to be preloaded

**Comment by axtloss** [comment]
It would be possible to just enable it by default and have it only preload the main executable (what wine/bottles executes), this would probably avoid taking up too much ram that could cause any issues.
Users with more ram could also enable to preload the whole directory where the application is, which would be more effective, but also take up more ram.

File: src/backend/wine/winecommand.py [hunk 2]
```diff
@@ -506,10 +507,43 @@ def __get_gamescope_cmd(self, return_steam_cmd: bool = False) -> str:
 
         return " ".join(gamescope_cmd)
 
+    def vmtouch_preload(self):
+        vmtouch_command = "/app/bin/vmtouch"
```
Comment (mirkobrombin) [inline src/backend/wine/winecommand.py:511]: I think shutil.which("vmtouch") better fit the case as we can also check it before executing. Also moving to globals.py like other checks, e.g.:
https://github.com/bottlesdevs/Bottles/blob/3ef4024fa2213abba405485aa5ac07fcff4c63b1/src/backend/globals.py#L95
and then check:
https://github.com/bottlesdevs/Bottles/blob/3ef4024fa2213abba405485aa5ac07fcff4c63b1/src/backend/wine/winecommand.py#L430

File: src/views/bottle_preferences.py [hunk 3]
```diff
@@ -55,6 +56,7 @@ class PreferencesView(Adw.PreferencesPage):
     btn_manage_gamescope = Gtk.Template.Child()
     btn_manage_sandbox = Gtk.Template.Child()
     btn_manage_versioning_patterns = Gtk.Template.Child()
+    btn_manage_vmtouch = Gtk.Template.Child()
```
Comment (mirkobrombin) [inline src/views/bottle_preferences.py:59]: You may want to make the switcher non sensitive if vmtouch is not present, just by using the check mentioned in the previous comment:

https://github.com/bottlesdevs/Bottles/blob/3ef4024fa2213abba405485aa5ac07fcff4c63b1/src/views/bottle_preferences.py#L213

**Review by mirkobrombin — COMMENTED** [review]
Just some improvements

File: src/views/bottle_preferences.py (hunk 3 above)
Comment (axtloss) [reply src/views/bottle_preferences.py:59]: But vmtouch would be installed in the flatpak since I added it to the manifest

Comment (axtloss) [reply src/views/bottle_preferences.py:59]: i just remembered that bottles also has packages other than the flatpak, I'll change the code according to your suggestion

File: com.usebottles.bottles.dev.yml [hunk 4]
```diff
@@ -189,6 +189,14 @@ modules:
...
   # Tools / Codecs
   # ----------------------------------------------------------------------------
+  - name: vmtouch
+    buildsystem: simple
+    sources:
+      - type: git
+        url: https://github.com/hoytech/vmtouch
+    build-commands:
+      - make
+      - make install PREFIX=/app
```
Comment (TheEvilSkeleton) [inline com.usebottles.bottles.dev.yml:199]: Don't forget to add this in the json manifest.

**Review by TheEvilSkeleton — CHANGES_REQUESTED** [review]

**Review by TheEvilSkeleton — APPROVED** [review]

**Comment by TheEvilSkeleton** [comment]
Marking it as draft to make it clearer for everyone.

**Comment by axtloss** [comment]
This isn't meant to be a draft anymore, i just forgot to remove the `[WIP]`

**Pull request merged**